### PR TITLE
 Improve actor performance via simplication

### PR DESCRIPTION
**Pull request by jhickson**
Changes related to performance:

- Removed state machine in `ActorImpl` and merged `ActorImpl` into `Actor`;
- Refined `ActorTaskScheduler` and the way the kind of task being executed is communicated so that using async state is no longer necessary;
- Replaced use of `Queue<>` in `ActorTaskScheduler` with something that easily allows putting something at the front, which simplified some logic;
- Made use of pause/resume functionality in `ActorTaskScheduler` to simplify starting an `Actor`;
- Made locking in `ActorTaskScheduler` more fine-grained;
- Used `UnsafeQueueUserWorkItem` on platforms that support it.

Also:

- Added NET Standard 2.0 support;
- Fixed various long-standing bugs that appeared in edge cases.

Overall, the changes made have reduced locking and complexity, which, apart from anything, has made it easier to track down long-standing spurious test failures and bugs. Performance isn't monumentally better but confidence that things are working correctly has been increased somewhat.

### CODE REVIEW

File: Winton.Extensions.Threading.Actor/Internal/ActorTaskScheduler.cs [hunk 1]
```diff
@@ -51,206 +53,261 @@ public ActorPauseAwaitable<T> WhileActorPaused<T>(Task<T> task)
...
+                        _back = actorTask;
+                        ReleaseLock();
+                    }
+
+                    break;
             }
         }
 
-        protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued)
+        private void TakeLock()
```
Comment (jhickson) [inline Winton.Extensions.Threading.Actor/Internal/ActorTaskScheduler.cs]: Note that the taking and releasing of the lock was isolated partly in order to experiment with using a `SpinLock`. However, using a `SpinLock` made no observable difference, so I've stuck with using a standard lock.

File: Winton.Extensions.Threading.Actor/Internal/ActorTask.cs [hunk 2]
```diff
@@ -1,24 +1,65 @@
...
+            Traits = _currentActorTaskTraits;
+            _canceller = _currentCanceller;
+
+            _currentCanceller = null;
+            _currentActorTaskTraits = ActorTaskTraits.None;
+        }
+
+        public static CancellationTokenSource CurrentCanceller
+        {
+            set => _currentCanceller = value;
```
[the remark this answers is not in the document]
Comment (jhickson) [reply Winton.Extensions.Threading.Actor/Internal/ActorTask.cs]: Correct - the `[ThreadStatic]` can only be applied to fields.

File: Winton.Extensions.Threading.Actor/Internal/ActorTask.cs [hunk 3]
```diff
@@ -1,24 +1,65 @@
...
+        public ActorTask(Task task)
+        {
+            Debug.Assert(_currentCanceller != null);
+
+            Task = task;
+            Traits = _currentActorTaskTraits;
+            _canceller = _currentCanceller;
+
+            _currentCanceller = null;
+            _currentActorTaskTraits = ActorTaskTraits.None;
```
[the remark this answers is not in the document]
Comment (jhickson) [reply Winton.Extensions.Threading.Actor/Internal/ActorTask.cs]: The values are only intended for the construction of a single `ActorTask` so I reset them after each construction. `ActorTaskFactory` sets these then queues a `Task` (via `Task.Factory.StartNew`) on the `ActorTaskScheduler` which will result in the synchronous call to `QueueTask` that constructs `ActorTask`.

It's a way of injecting data the scheduler needs. Previously I used `Task.AsyncState` for this but this is more performant for various reasons.

[the remark this answers is not in the document]
Comment (Choc13) [reply Winton.Extensions.Threading.Actor/Internal/ActorTask.cs]: From my naive understanding, would it not be possible to pass those values to the `ActorTask` constructor from the `ActorTaskFactory`? That would avoid the shared construction state across different instances.

[the remark this answers is not in the document]
Comment (jhickson) [reply Winton.Extensions.Threading.Actor/Internal/ActorTask.cs]: No. To do that I would need `ActorTask` to derive from `Task` so that in `ActorTaskFactory` I could do:

```
var actorTask = new ActorTask(...);
actorTask.Start(_scheduler);
```

rather than calling `Task.Factory.StartNew`. However, directly instantiating a `Task` then starting it rather than calling `Task.Run` or `Task.Factory.StartNew` is a bad idea (there's a [Stephen Cleary blog](http://blog.stephencleary.com/2014/05/a-tour-of-task-part-1-constructors.html) about it). You can get race conditions with the passed-in cancellation token being cancelled before `Start` is called on the `Task`, so that `Start` will throw an `InvalidOperationException`. I actually went this route initially and hit this race in tests. You can try to come up with a mechanism to get round it but it's nasty and not fool-proof. This injection via thread statics is the simplest alternative given I therefore need to use a vanilla `Task` outside the scheduler.
